### PR TITLE
CNV-13710: Configuring cluster for real-time workloads

### PR DESCRIPTION
**Pull request by sjhala-ccs**
Version(s): 4.16+

Issue: [CNV-13710](https://issues.redhat.com//browse/CNV-13710)

Link to docs preview: https://74365--ocpdocs-pr.netlify.app/openshift-enterprise/latest/virt/virtual_machines/advanced_vm_management/virt-configuring-cluster-realtime-workloads.html

QE review:
- [x] QE has approved this change.

Additional information:

### CODE REVIEW

File: modules/virt-configuring-vm-real-time.adoc [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+      autoattachGraphicsDevice: false
+      autoattachMemBalloon: false
+      autoattachSerialConsole: true
+    ioThreadsPolicy: auto
+    memory:
+      guest: 4Gi
+      hugepages:
+        pageSize: 1Gi <6>
+  terminationGracePeriodSeconds: 0
+# ...
```
Comment (orelmisan) [inline modules/virt-configuring-vm-real-time.adoc:50]:
```suggestion
template:
  annotations:
    cpu-load-balancing.crio.io: disable <1>
    cpu-quota.crio.io: disable <2>
    irq-load-balancing.crio.io: disable <3>
  spec:
    domain:
    cpu:
      dedicatedCpuPlacement: true
      isolateEmulatorThread: true
      model: host-passthrough
      numa:
        guestMappingPassthrough: {}
      realtime: {}
      sockets: 1 <4>
      cores: 4 <5>
      threads: 1
    devices:
      autoattachGraphicsDevice: false
      autoattachMemBalloon: false
      autoattachSerialConsole: true
    ioThreadsPolicy: auto
    memory:
      guest: 4Gi
      hugepages:
       pageSize: 1Gi <6>
   terminationGracePeriodSeconds: 0
```

**Review by orelmisan — COMMENTED** [review]
Thank you for the PR @sjhala-ccs 
Please see the inline comments.

**Review by RamLavi — CHANGES_REQUESTED** [review]

File: modules/virt-configuring-cluster-real-time.adoc [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+You can configure an {product-title} cluster to run real-time workloads.
+
+.Prerequisites
+* You have access to the cluster as a user with `cluster-admin` permissions.
+* You have installed the OpenShift CLI (`oc`).
+* You have installed the Node Tuning Operator.
+
+.Procedure
+
+. Label a subset of the compute nodes with a custom role; for example, `worker-realtime`:
```
[the remark this answers is not in the document]
Comment (orelmisan) [reply modules/virt-configuring-cluster-real-time.adoc:18]: Interesting point.
Maybe its worth adding it as a note.

File: modules/virt-configuring-cluster-real-time.adoc [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  name: profile-1
+spec:
+  cpu:
+    isolated: 4-39,44-79
+    reserved: 0-3,40-43
+  globallyDisableIrqLoadBalancing: true
+  hugepages:
+    defaultHugepagesSize: 1G
+    pages:
+    - count: 16
```
[the remark this answers is not in the document]
Comment (orelmisan) [reply modules/virt-configuring-cluster-real-time.adoc:66]: Currently, real-time VMs in KubeVirt require `spec.Domain.CPU.NUMA.GuestMappingPassthrough` which in turn require hugepages.
There is a [discussion](https://groups.google.com/g/kubevirt-dev/c/8z4xFZ7czo4) about it in the kubevirt-dev mailing list.

File: modules/virt-configuring-vm-real-time.adoc [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+----
+$ oc apply -f <file_name>.yaml
+----
+
+. Configure the guest operating system. The following example shows the configuration steps for {op-system-base} 8 OS:
+.. Configure huge pages by using the GRUB bootloader command-line interface. In the following example, 8 1G huge pages are specified.
++
+[source,terminal]
+----
+$ grubby --update-kernel=ALL --args="default_hugepagesz=1GB hugepagesz=1G hugepages=8"
```
[the remark this answers is not in the document]
Comment (orelmisan) [reply modules/virt-configuring-vm-real-time.adoc:72]: @RamLavi Please see my [comment](https://github.com/openshift/openshift-docs/pull/74365#discussion_r1557020243).

File: modules/virt-configuring-cluster-real-time.adoc [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+* You have installed the OpenShift CLI (`oc`).
+* You have installed the Node Tuning Operator.
+
+.Procedure
+
+. Label a subset of the compute nodes with a custom role; for example, `worker-realtime`:
++
+[source,terminal]
+----
+$ oc label node <node_name> node-role.kubernetes.io/worker-realtime=""
```
Comment (segevyos) [inline modules/virt-configuring-cluster-real-time.adoc:22]: This will not work in compact/SNO clusters (I don't remember why, and how to find the ticket that details it). For those, the configuration will rely on the already-existing `master` label and MCP.

File: modules/virt-configuring-cluster-real-time.adoc [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+----
+
+. Create a new `MachineConfigPool` manifest that contains the `worker-realtime` label in the `spec.machineConfigSelector` object:
```
Comment (segevyos) [inline modules/virt-configuring-cluster-real-time.adoc:25]: Again - not relevant for compact and SNO clusters.

File: modules/virt-configuring-cluster-real-time.adoc (hunk 3 above)
[the remark this answers is not in the document]
Comment (segevyos) [reply modules/virt-configuring-cluster-real-time.adoc:66]: From what I see in my env, 8 is indeed enough, as @RamLavi suggests.
And we also had a [discussion about it in slack](https://redhat-internal.slack.com/archives/C05K99TGQ14/p1710860963629769), where we said that 4 would probably be enough, and we agreed that 8 is the safest choice.

File: modules/virt-configuring-cluster-real-time.adoc [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    realTime: true
+  nodeSelector:
+    node-role.kubernetes.io/worker-realtime: ""
+  numa:
+    topologyPolicy: single-numa-node
+----
++
+[NOTE]
+====
+The compute nodes automatically restart after you apply the `MachineConfigPool` and `PerformanceProfile` manifests.
```
Comment (segevyos) [inline modules/virt-configuring-cluster-real-time.adoc:81]: AFAIU each node is rebooted **twice**. The user should be aware of that, so they don't suspect something is wrong if they see nodes restarted again, and the entire process taking quite a long time.

[the remark this answers is not in the document]
Comment (segevyos) [reply modules/virt-configuring-cluster-real-time.adoc:81]: +1

File: modules/virt-configuring-vm-real-time.adoc [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+. Save and exit the editor.
+. Apply the `VirtualMachine` manifest:
++
+[source,terminal]
+----
+$ oc apply -f <file_name>.yaml
+----
+
+. Configure the guest operating system. The following example shows the configuration steps for {op-system-base} 8 OS:
```
Comment (segevyos) [inline modules/virt-configuring-vm-real-time.adoc:67]: If this should be set on the guest VM, then the user should first be instructed to access the VM (`virtctl console <vm>`).
This also means adding the installation of `virtctl` as a prerequisite.

**Review by segevyos — CHANGES_REQUESTED** [review]

File: modules/virt-configuring-cluster-real-time.adoc (hunk 7 above)
[the remark this answers is not in the document]
Comment (RamLavi) [reply modules/virt-configuring-cluster-real-time.adoc:81]: @orelmisan yeah your suggestion is great IMO.

File: modules/virt-configuring-vm-real-time.adoc (hunk 4 above)
[the remark this answers is not in the document]
Comment (RamLavi) [reply modules/virt-configuring-vm-real-time.adoc:72]: are you sure the comment link is correct?

[the remark this answers is not in the document]
Comment (orelmisan) [reply modules/virt-configuring-vm-real-time.adoc:72]: Yes, my agreement with @sjhala-ccs was that this part will be copy-pasted from the DPDK docs, and we will revisit it in a follow-up PR after we will have all the data.

[the remark this answers is not in the document]
Comment (RamLavi) [reply modules/virt-configuring-vm-real-time.adoc:72]: ah so right not we're just copy pasting? 
Then I don't understand what's the purpose of this review

[the remark this answers is not in the document]
Comment (orelmisan) [reply modules/virt-configuring-vm-real-time.adoc:72]: Only the part of guest configuration is copy-pasted.
The rest should be thoroughly reviewed.

File: modules/virt-configuring-vm-real-time.adoc [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+          hugepages:
+            pageSize: 1Gi <6>
+        terminationGracePeriodSeconds: 0
+# ...
+----
+<1> This annotation specifies that load balancing is disabled for CPUs that are used by the container.
+<2> This annotation specifies that the CPU quota is disabled for CPUs that are used by the container.
+<3> This annotation specifies that Interrupt Request (IRQ) load balancing is disabled for CPUs that are used by the container.
+<4> The number of sockets inside the VM. This field must be set to `1` for the CPUs to be scheduled from the same Non-Uniform Memory Access (NUMA) node.
+<5> The number of cores inside the VM. This must be a value greater than or equal to `1`. In this example, the VM is scheduled with 5 hyper-threads or 10 CPUs.
```
Comment (orelmisan) [inline modules/virt-configuring-vm-real-time.adoc:60]:
```suggestion
<5> The number of cores inside the VM. This must be a value greater than or equal to `1`.
```

File: modules/virt-configuring-vm-real-time.adoc [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+          guest: 4Gi
+          hugepages:
+            pageSize: 1Gi <6>
+        terminationGracePeriodSeconds: 0
+# ...
+----
+<1> This annotation specifies that load balancing is disabled for CPUs that are used by the container.
+<2> This annotation specifies that the CPU quota is disabled for CPUs that are used by the container.
+<3> This annotation specifies that Interrupt Request (IRQ) load balancing is disabled for CPUs that are used by the container.
+<4> The number of sockets inside the VM. This field must be set to `1` for the CPUs to be scheduled from the same Non-Uniform Memory Access (NUMA) node.
```
Comment (orelmisan) [inline modules/virt-configuring-vm-real-time.adoc:59]: >  This field must be set to `1` for the CPUs to be scheduled from the same Non-Uniform Memory Access (NUMA) node.

I'm not sure that this statement is true.
I suggest to remove this part.

**Review by orelmisan — COMMENTED** [review]
Thank you @sjhala-ccs, overall looks good.
Just a few small comments.

@matosatti could you please have a look?
Please ignore the guest OS configuration part as it would be revisited in a follow-up PR.

**Comment by sjhala-ccs** [comment]
> Please ignore the guest OS configuration part as it would be revisited in a follow-up PR.

@orelmisan Is there a separate Jira where this effort can be tracked?

**Review by orelmisan — APPROVED** [review]
Thank you

**Comment by orelmisan** [comment]
> > Please ignore the guest OS configuration part as it would be revisited in a follow-up PR.
> 
> @orelmisan Is there a separate Jira where this effort can be tracked?

@RamLavi and I, will create the follow-up PR to replace this part:
https://issues.redhat.com/browse/CNV-40785

File: modules/virt-configuring-cluster-real-time.adoc [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+. Label a subset of the compute nodes with a custom role; for example, `worker-realtime`:
++
+[source,terminal]
+----
+$ oc label node <node_name> node-role.kubernetes.io/worker-realtime=""
+----
++
+[NOTE]
+====
+You must use the default `master` role for {sno} clusters.
```
Comment (segevyos) [inline modules/virt-configuring-cluster-real-time.adoc:27]: Compact clusters as well (not only SNO).

File: modules/virt-configuring-cluster-real-time.adoc [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+          - worker
+          - worker-realtime
+  nodeSelector:
+    matchLabels:
+      node-role.kubernetes.io/worker-realtime: "" 
+----
++
+[NOTE]
+====
+The above example manifest does not apply to {sno} clusters.
```
Comment (segevyos) [inline modules/virt-configuring-cluster-real-time.adoc:56]: @orelmisan @RamLavi 
I'm not sure if this is true (even though I commented about it in my previous review), because now I see that this is the same MCP I applied successfully (i.e. it ended with successful checkups) on my compact cluster.
WDYT?

File: modules/virt-configuring-cluster-real-time.adoc (hunk 7 above)
Comment (segevyos) [reply modules/virt-configuring-cluster-real-time.adoc:91]: I don't see any reference to this.

**Review by segevyos — CHANGES_REQUESTED** [review]

File: modules/virt-configuring-vm-real-time.adoc [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+
+. Configure the guest operating system. The following example shows the configuration steps for {op-system-base} 8 operating system:
+.. Run the following command to connect to the VM console:
++
+[source,terminal]
+----
+$ virtctl console <vm_name>
+----
+
+.. Configure huge pages by using the GRUB boot loader command-line interface. In the following example, 8 1G huge pages are specified.
```
Comment (bergerhoffer) [inline modules/virt-configuring-vm-real-time.adoc:79]: Again, should normally be a space between the value and the units, but I believe that if there is a number before it as well, then you should use a hyphen. You could probably also spell it out too as an alternative.

So either: "eight 1 G huge pages" or "8 1-G huge pages"

File: modules/virt-configuring-vm-real-time.adoc [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+[source,terminal]
+----
+$ dnf install -y tuned-profiles-cpu-partitioning
+----
++
+[source,terminal]
+----
+$ echo isolated_cores=2-9 > /etc/tuned/cpu-partitioning-variables.conf
+----
+The first two CPUs (0 and 1) are set aside for house keeping tasks and the rest are isolated for the real-time application.
```
Comment (bergerhoffer) [inline modules/virt-configuring-vm-real-time.adoc:97]: Just wondering if "housekeeping tasks" is too colloquial. Is there a better term we could use instead here?

**Review by bergerhoffer — COMMENTED** [review]
Nice job, a few comments/suggestions!

File: modules/virt-configuring-vm-real-time.adoc [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+        memory:
+          guest: 4Gi
+          hugepages:
+            pageSize: 1Gi <6>
+        terminationGracePeriodSeconds: 0
+# ...
+----
+<1> This annotation specifies that load balancing is disabled for CPUs that are used by the container.
+<2> This annotation specifies that the CPU quota is disabled for CPUs that are used by the container.
+<3> This annotation specifies that Interrupt Request (IRQ) load balancing is disabled for CPUs that are used by the container.
```
[the remark this answers is not in the document]
Comment (sjhala-ccs) [reply modules/virt-configuring-vm-real-time.adoc:58]: Changing to lowercase

File: modules/virt-configuring-cluster-real-time.adoc [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+  nodeSelector:
+    matchLabels:
+      node-role.kubernetes.io/worker-realtime: "" 
+----
++
+[NOTE]
+====
+You do not need to create a new `MachineConfigPool` manifest for {sno} and compact clusters.
+====
+
```
[the remark this answers is not in the document]
Comment (sjhala-ccs) [reply modules/virt-configuring-cluster-real-time.adoc:58]: The note after step 3 implies that the user needs to apply these resources, but I can add the extra steps. I feel, that the usage across our docs is not consistent.

File: modules/virt-configuring-cluster-real-time.adoc [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+          - worker
+          - worker-realtime
+  nodeSelector:
+    matchLabels:
+      node-role.kubernetes.io/worker-realtime: "" 
+----
++
+[NOTE]
+====
+You do not need to create a new `MachineConfigPool` manifest for {sno} and compact clusters.
```
[the remark this answers is not in the document]
Comment (sjhala-ccs) [reply modules/virt-configuring-cluster-real-time.adoc:56]: Leaving this as is

File: modules/virt-configuring-cluster-real-time.adoc [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+$ oc patch hyperconverged kubevirt-hyperconverged -n {CNVNamespace} \
+    --type='json' -p='[{"op": "add", "path": "/spec/defaultRuntimeClass", "value":"<runtimeclass-name>"}]'
+----
++
+[NOTE]
+====
+Editing the `HyperConverged` CR changes a global setting that affects all VMs that are created after the change is applied.
+====
+
+. If your real-time-enabled compute nodes use Simultaneous multithreading (SMT), enable the `AlignCPUs` enabler by editing the `HyperConverged` CR:
```
[the remark this answers is not in the document]
Comment (sjhala-ccs) [reply modules/virt-configuring-cluster-real-time.adoc:114]: Changed to lowercase

File: modules/virt-configuring-vm-real-time.adoc [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+= Configuring a virtual machine for real-time workloads
+
+You can run real-time workloads on virtual machines (VMs).
+
+.Prerequisites
+* Your cluster is configured to run real-time workloads.
+* You have installed the `virtctl` tool.
+
+.Procedure
+. Edit the `VirtualMachine` manifest to include information about CPU topology, CRI-O annotations, and huge pages:
```
[the remark this answers is not in the document]
Comment (sjhala-ccs) [reply modules/virt-configuring-vm-real-time.adoc:16]: I think we are creating a new manifest. I will update the wording to reflect that.

File: modules/virt-configuring-vm-real-time.adoc [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+----
+<1> This annotation specifies that load balancing is disabled for CPUs that are used by the container.
+<2> This annotation specifies that the CPU quota is disabled for CPUs that are used by the container.
+<3> This annotation specifies that Interrupt Request (IRQ) load balancing is disabled for CPUs that are used by the container.
+<4> The number of sockets inside the VM. 
+<5> The number of cores inside the VM. This must be a value greater than or equal to `1`.
+<6> The size of the huge pages. The possible values for x86-64 architecture are 1Gi and 2Mi. In this example, the request is for 4 huge pages of size 1Gi.
+
+. Save and exit the editor.
+. Apply the `VirtualMachine` manifest:
```
[the remark this answers is not in the document]
Comment (sjhala-ccs) [reply modules/virt-configuring-vm-real-time.adoc:64]: See my comment [here](https://github.com/openshift/openshift-docs/pull/74365#discussion_r1575344680)

File: modules/virt-configuring-vm-real-time.adoc (hunk 14 above)
Comment (sjhala-ccs) [reply modules/virt-configuring-vm-real-time.adoc:97]: It is a [term used in computing](https://en.wikipedia.org/wiki/Housekeeping_(computing)) as well. Do you have any other suggestions?

Comment (bergerhoffer) [reply modules/virt-configuring-vm-real-time.adoc:96]: Fair enough then!

**Comment by matosatti** [comment]
List of items that are in the https://access.redhat.com/solutions/7007632 document,
but not on the DPDK document
(https://74365--ocpdocs-pr.netlify.app/openshift-enterprise/latest/virt/vm_networking/virt-using-dpdk-with-sriov)
and not on this document:

1) Configure CPU Manager with the static cpuManagerPolicy.

2) Configure Memory Manager to ensure VM memory is affinitized to the proper NUMA node where the vcpus are executing.

3) No explanation for the following options (why not? don't
see any benefit of not including the explanations):

dedicatedCpuPlacement has to be true to indicate the desire of allocating dedicated CPU resources.
Set isolateEmulatorThread to true to use a different CPU for the qemu emulator thread.
Set ioThreadsPolicy to auto to use different threads for the disk IO.
Set cpu-quota.crio.io: "disable" in the pod annotation. Otherwise CPU CFS quota will be enabled for the pod,
causing the scheduler to throttle the vcpu thread. Unfortunately, using the cpu-quota.crio.io pod annotation requires setting a runtime class, as documented in the "Disabling CPU CFS quota" section of the OpenShift documentation. But it is not possible to set runtime classes for VM pods (see https://bugzilla.redhat.com/show_bug.cgi?id=2203291), so the following workaround is necessary to set runtime class of all VMs in the cluster:

Items that are necessary but not on the https://access.redhat.com/solutions/7007632
===================================================================================
document:
=========

4) Reducing NIC queues

From https://74365--ocpdocs-pr.netlify.app/openshift-enterprise/latest/scalability_and_performance/cnf-low-latency-tuning#node-tuning-operator-known-limitations-for-real-time_cnf-master:

Reducing NIC queues using the Node Tuning Operator
The Node Tuning Operator allows you to adjust the network interface controller (NIC) queue count for each network device. By using a PerformanceProfile, the amount of queues can be reduced to the number of reserved CPUs.

Specific to realtime configuration (this PR):
============================================

5) Using tuned-profiles-cpu-partitioning instead of realtime-virtual-guest.
Should mention realtime-virtual-guest, not cpu-partitioning profile.

Replace the following steps:
===========================

To achieve low-latency tuning by using the cpu-partitioning profile in the TuneD application, run the following commands:


$ dnf install -y tuned-profiles-cpu-partitioning

$ echo isolated_cores=2-9 > /etc/tuned/cpu-partitioning-variables.conf
The first two CPUs (0 and 1) are set aside for house keeping tasks and the rest are isolated for the real-time application.

$ tuned-adm profile cpu-partitioning

With:
=====

$ Install realtime-virtual-guest profile (https://access.redhat.com/solutions/6923381)

$ Edit /etc/tuned/realtime-virtual-guest-variables.conf

$ Adjust the isolated_cores variable, for example:

isolated_cores=2-9

$ Adjust the netdev_queue_count, set it to the number
of housekeeping CPUs:

# Set the desired combined queue count value using the parameter provided
# below. Ideally this should be set to the number of housekeeping CPUs i.e.,
# in the example given below it is assumed that the system has 4 housekeeping
# (non-isolated) CPUs.
#
# netdev_queue_count=4

$ reboot the guest.


6) Please add a pointer to the host low latency configuration:

https://74365--ocpdocs-pr.netlify.app/openshift-enterprise/latest/scalability_and_performance/cnf-low-latency-tuning

**Comment by orelmisan** [comment]
@matosatti lets discuss it offline.

**Comment by orelmisan** [comment]
> List of items that are in the https://access.redhat.com/solutions/7007632 document, but not on the DPDK document (https://74365--ocpdocs-pr.netlify.app/openshift-enterprise/latest/virt/vm_networking/virt-using-dpdk-with-sriov) and not on this document:
> 
> 1. Configure CPU Manager with the static cpuManagerPolicy.
> 2. Configure Memory Manager to ensure VM memory is affinitized to the proper NUMA node where the vcpus are executing.

This is performed automatically when the PerformanceProfile is applied.

> 3. No explanation for the following options (why not? don't
>    see any benefit of not including the explanations):
> 
> dedicatedCpuPlacement has to be true to indicate the desire of allocating dedicated CPU resources. Set isolateEmulatorThread to true to use a different CPU for the qemu emulator thread. Set ioThreadsPolicy to auto to use different threads for the disk IO. Set cpu-quota.crio.io: "disable" in the pod annotation. Otherwise CPU CFS quota will be enabled for the pod, causing the scheduler to throttle the vcpu thread. Unfortunately, using the cpu-quota.crio.io pod annotation requires setting a runtime class, as documented in the "Disabling CPU CFS quota" section of the OpenShift documentation. But it is not possible to set runtime classes for VM pods (see https://bugzilla.redhat.com/show_bug.cgi?id=2203291), so the following workaround is necessary to set runtime class of all VMs in the cluster:

These explanations are present bellow the VM manifest.
What do you think is missing?

> # Items that are necessary but not on the https://access.redhat.com/solutions/7007632
> # document:
> 4. Reducing NIC queues
> 
> From https://74365--ocpdocs-pr.netlify.app/openshift-enterprise/latest/scalability_and_performance/cnf-low-latency-tuning#node-tuning-operator-known-limitations-for-real-time_cnf-master:
> 
> Reducing NIC queues using the Node Tuning Operator The Node Tuning Operator allows you to adjust the network interface controller (NIC) queue count for each network device. By using a PerformanceProfile, the amount of queues can be reduced to the number of reserved CPUs.

The example VM does not contain a NIC.
Are you referring to the host's NIC or the guest's?

> # Specific to realtime configuration (this PR):
> 5. Using tuned-profiles-cpu-partitioning instead of realtime-virtual-guest.
>    Should mention realtime-virtual-guest, not cpu-partitioning profile.
> 
> # Replace the following steps:
> To achieve low-latency tuning by using the cpu-partitioning profile in the TuneD application, run the following commands:
> 
> $ dnf install -y tuned-profiles-cpu-partitioning
> 
> $ echo isolated_cores=2-9 > /etc/tuned/cpu-partitioning-variables.conf The first two CPUs (0 and 1) are set aside for house keeping tasks and the rest are isolated for the real-time application.
> 
> $ tuned-adm profile cpu-partitioning
> 
> # With:
> $ Install realtime-virtual-guest profile (https://access.redhat.com/solutions/6923381)
> 
> $ Edit /etc/tuned/realtime-virtual-guest-variables.conf
> 
> $ Adjust the isolated_cores variable, for example:
> 
> isolated_cores=2-9
> 
> $ Adjust the netdev_queue_count, set it to the number of housekeeping CPUs:
> 
> # Set the desired combined queue count value using the parameter provided
> # below. Ideally this should be set to the number of housekeeping CPUs i.e.,
> # in the example given below it is assumed that the system has 4 housekeeping
> # (non-isolated) CPUs.
> # netdev_queue_count=4
> $ reboot the guest.
> 
> 6. Please add a pointer to the host low latency configuration:
> 
> https://74365--ocpdocs-pr.netlify.app/openshift-enterprise/latest/scalability_and_performance/cnf-low-latency-tuning

@RamLavi and I will add the guest configuration steps in a follow up.
The information presented here is relevant to DPDK machine and is here so we will have a structure for the follow-up.

> 7. What about steps to test latency inside the VM ?

Could you please provide them?
Is it something that we can do in a follow-up in order to move forward with this PR?

> 8. After this is addressed, please show that steps have been tested by QE.

@segevyos and Nini Gu (I'm not sure what her GitHub handle is) went over this document.

File: modules/virt-configuring-cluster-real-time.adoc [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+. Apply the `PerformanceProfile` manifest:
++
+[source,terminal]
+----
+$ oc apply -f <real_time_pp>.yaml
+----
++
+[NOTE]
+====
+The compute nodes automatically reboot twice after you apply the `MachineConfigPool` and `PerformanceProfile` manifests. This process might take a few minutes.
```
Comment (segevyos) [inline modules/virt-configuring-cluster-real-time.adoc:105]: It takes much more than a few minutes, so this might be confusing, and lead the user to assume that if the process is not completed after 20-30 minutes (and it probably won't) then there's something wrong.
I would rephrase it to something that reflects the fact the even more than a 1 hour wait if possible.

File: modules/virt-configuring-vm-real-time.adoc [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+
+. Configure the guest operating system. The following example shows the configuration steps for a {op-system-base} 8 operating system:
+.. Run the following command to connect to the VM console:
++
+[source,terminal]
+----
+$ virtctl console <vm_name>
+----
+
+.. Configure huge pages by using the GRUB boot loader command-line interface. In the following example, 8 1-G huge pages are specified.
```
Comment (segevyos) [inline modules/virt-configuring-vm-real-time.adoc:78]: > 1-G huge pages

Are you sure that using the hyphen is a valid tech syntax? I'm asking because I have never encountered this before, so I apologize if I'm just making unnecessary noise here.

**Review by segevyos — CHANGES_REQUESTED** [review]

**Comment by matosatti** [comment]
> > List of items that are in the https://access.redhat.com/solutions/7007632 document, but not on the DPDK document (https://74365--ocpdocs-pr.netlify.app/openshift-enterprise/latest/virt/vm_networking/virt-using-dpdk-with-sriov) and not on this document:
> > 
> > 1. Configure CPU Manager with the static cpuManagerPolicy.
> > 2. Configure Memory Manager to ensure VM memory is affinitized to the proper NUMA node where the vcpus are executing.
> 
> This is performed automatically when the PerformanceProfile is applied.

No, its not: please mention the documentation below:

See "Memory Manager" section from:

https://access.redhat.com/articles/6994974

Memory Manager
When VMs are small enough to fit within a NUMA node, Memory Manager can be used in conjunction with the Topology Manager policies of single-numa-node or restricted to ensure memory is also affinitized to a preferred NUMA node. See the NUMA-aware scheduling section for more information.

If a VM is large enough to span multiple NUMA nodes on a system, the memory affinitization can be controlled through a VM option called guestMappingPassthrough instead, see the [VM Tuning]{#vm-config-tuning} section for more information.

Memory Manager can be enabled by configuring the “static” policy using a kubeletConfig similar to the CPU and Topology Managers. Configuring Memory Manager also requires setting a reservedMemory value for at least one NUMA node. This value should equal the sum of all reserved memory for a node (kube-reserved + system-reserved + eviction-threshold). This value can be calculated by comparing a node’s total reserved value: Memory Capacity - Memory Allocatable. Example kubeletConfig syntax is provided below:

> > 3. No explanation for the following options (why not? don't
> >    see any benefit of not including the explanations):
> > 
> > dedicatedCpuPlacement has to be true to indicate the desire of allocating dedicated CPU resources. Set isolateEmulatorThread to true to use a different CPU for the qemu emulator thread. Set ioThreadsPolicy to auto to use different threads for the disk IO. Set cpu-quota.crio.io: "disable" in the pod annotation. Otherwise CPU CFS quota will be enabled for the pod, causing the scheduler to throttle the vcpu thread. Unfortunately, using the cpu-quota.crio.io pod annotation requires setting a runtime class, as documented in the "Disabling CPU CFS quota" section of the OpenShift documentation. But it is not possible to set runtime classes for VM pods (see https://bugzilla.redhat.com/show_bug.cgi?id=2203291), so the following workaround is necessary to set runtime class of all VMs in the cluster:
> 
> These explanations are present bellow the VM manifest. What do you think is missing?

The documentation in the VM manifest just repeats whats signified by the variable name, for example:

        cpu-load-balancing.crio.io: disable 

Can you at least include links to the proper documentation which explains the flags in more depth?
> 
> > # Items that are necessary but not on the https://access.redhat.com/solutions/7007632
> > # document:
> > 
> > 4. Reducing NIC queues
> > 
> > From https://74365--ocpdocs-pr.netlify.app/openshift-enterprise/latest/scalability_and_performance/cnf-low-latency-tuning#node-tuning-operator-known-limitations-for-real-time_cnf-master:
> > Reducing NIC queues using the Node Tuning Operator The Node Tuning Operator allows you to adjust the network interface controller (NIC) queue count for each network device. By using a PerformanceProfile, the amount of queues can be reduced to the number of reserved CPUs.
> 
> The example VM does not contain a NIC. 

It probably should, because i don't see a customer using a VM without a NIC.

> Are you referring to the host's NIC or the guest's?

Hosts NIC's.

> > # Specific to realtime configuration (this PR):
> > 
> > 5. Using tuned-profiles-cpu-partitioning instead of realtime-virtual-guest.
> >    Should mention realtime-virtual-guest, not cpu-partitioning profile.
> > 
> > # Replace the following steps:
> > To achieve low-latency tuning by using the cpu-partitioning profile in the TuneD application, run the following commands:
> > $ dnf install -y tuned-profiles-cpu-partitioning
> > $ echo isolated_cores=2-9 > /etc/tuned/cpu-partitioning-variables.conf The first two CPUs (0 and 1) are set aside for house keeping tasks and the rest are isolated for the real-time application.
> > $ tuned-adm profile cpu-partitioning
> > # With:
> > $ Install realtime-virtual-guest profile (https://access.redhat.com/solutions/6923381)
> > $ Edit /etc/tuned/realtime-virtual-guest-variables.conf
> > $ Adjust the isolated_cores variable, for example:
> > isolated_cores=2-9
> > $ Adjust the netdev_queue_count, set it to the number of housekeeping CPUs:
> > # Set the desired combined queue count value using the parameter provided
> > # below. Ideally this should be set to the number of housekeeping CPUs i.e.,
> > # in the example given below it is assumed that the system has 4 housekeeping
> > # (non-isolated) CPUs.
> > # netdev_queue_count=4
> > $ reboot the guest.
> > 
> > 6. Please add a pointer to the host low latency configuration:
> > 
> > https://74365--ocpdocs-pr.netlify.app/openshift-enterprise/latest/scalability_and_performance/cnf-low-latency-tuning
> 
> @RamLavi and I will add the guest configuration steps in a follow up. The information presented here is relevant to DPDK machine and is here so we will have a structure for the follow-up.
> 
> > 7. What about steps to test latency inside the VM ?
> 
> Could you please provide them? Is it something that we can do in a follow-up in order to move forward with this PR?

I think its better to have a full guide (similarly to the baremetal low latency configuration), rather than separate disconnected pieces (which makes it hard to test them as a single unit).

> > 8. After this is addressed, please show that steps have been tested by QE.
> 
> @segevyos and Nini Gu (I'm not sure what her GitHub handle is) went over this document.

Due to the large number of steps involved, i don't think a simple review is sufficient, but actually testing the steps and confirming they provide the necessary end result is desired.

**Comment by orelmisan** [comment]
@matosatti and I had an offline discussion.
@matosatti requested that a link to https://docs.openshift.com/container-platform/4.16/scalability_and_performance/cnf-low-latency-tuning.html#reducing-nic-queues-using-the-node-tuning-operator_cnf-master will be added.

The guest configuration will be revisited in a follow-up.

File: modules/virt-configuring-vm-real-time.adoc (hunk 22 above)
Comment (sjhala-ccs) [reply modules/virt-configuring-vm-real-time.adoc:78]: Removed the hyphen

File: modules/virt-configuring-cluster-real-time.adoc (hunk 21 above)
Comment (sjhala-ccs) [reply modules/virt-configuring-cluster-real-time.adoc:105]: Changed to "This process might take a long time".

**Review by orelmisan — APPROVED** [review]
Thank you
